### PR TITLE
refactor: collect executor and decoder states into dataclass

### PR DESCRIPTION
**Pull request by tongyuantongyu**

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 1]
```diff
@@ -923,37 +931,36 @@ def _executor_loop_overlap(self):
                         if r.get_context_remaining_length() == 0
                     ]
 
-                    self.previous_batch = (scheduled_batch, new_tensors_device,
-                                           new_tensors_host, decoder_event,
-                                           iter_start_time, iter_stats,
-                                           ctx_transmission_reqs)
-                    self._gather_dp_requests_num()
+                self.previous_batch = BatchState(
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: This line is moved from the scope of `if scheduled_batch.batch_size > 0:`. Is it expected?

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 2]
```diff
@@ -240,153 +225,162 @@ def _handle_stop_criteria(self, request: LlmRequest, new_token: int,
...
-        new_tokens_list = new_tensors_host["new_tokens_host"].tolist()
+    def update_requests(self, decoder_state: DecoderState) -> None:
+        if decoder_state.decoder_event:
+            decoder_state.decoder_event.synchronize()
+        new_tokens_list = decoder_state.new_tensors_host[
+            "new_tokens_host"].tolist()
 
-        idx = 0
+        request_idx = 0
+        logits_idx = 0
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: How about renaming `logits_idx` to `token_idx`? It's used to find a token in `new_tokens_list`.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 1 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Nice catch! Fixed.

File: tensorrt_llm/_torch/pyexecutor/decoder.py [hunk 3]
```diff
@@ -240,153 +225,162 @@ def _handle_stop_criteria(self, request: LlmRequest, new_token: int,
...
+                continue
 
-                # Accept draft tokens (if we have any) if and only if they match the new
-                # token exactly.
+            accepted_tokens, total_tokens = 1, 1
+            new_token = new_tokens_list[logits_idx]
+            num_tokens = request.add_new_token(new_token, beam_idx)
+            self._handle_stop_criteria(request, new_token, num_tokens, beam_idx)
+
+            if request.py_draft_tokens is not None:
```
Comment (QiJune) [inline tensorrt_llm/_torch/pyexecutor/decoder.py]: `request.py_draft_tokens is None` is called twice, one time in `sorted`, one time here. In the original implementation, it's only called one time. 

I suggest we keep the original implemention. For duplicate codes in `extend_requests` and `generation_requests`, we can write a helper function:

```python
        def _handle_request(request):     
            if request.state != LlmRequestState.GENERATION_COMPLETE:
                new_token = new_tokens_list[idx]
                num_tokens = request.add_new_token(new_token, beam_idx)
                self._handle_stop_criteria(request, new_token, num_tokens,
                                           beam_idx)
```

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 2 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: I reverted it back to `idx` in this PR. The rename is only reasonable with the addition of logits handling code.

File: tensorrt_llm/_torch/pyexecutor/decoder.py (hunk 3 above)
Comment (tongyuantongyu) [reply tensorrt_llm/_torch/pyexecutor/decoder.py]: I reverted most of the changes to `update_requests()`. They are not very meaningful without logits handling code.

Now the only changes are simplify skipping `chunked_requests` and remove `hasattr(request, "py_draft_tokens")` check. `py_draft_tokens` is always set by `LlmRequest.__init__()` now.

**Review by QiJune — APPROVED** [review]
LGTM

**Comment by tongyuantongyu** [comment]
@QiJune I realized I also need to do some refactoring to the newly added pipeline parallel variant of the executor loop functions. CC @amukkara
